blockchain/v0/reactor.go

```diff
@@ -391,71 +522,75 @@ FOR_LOOP:
 			if err != nil {
 				err = fmt.Errorf("invalid last commit: %w", err)
-				bcR.Logger.Error(err.Error(),
-					"last_commit", second.LastCommit, "block_id", firstID, "height", first.Height)
-
-				peerID := bcR.pool.RedoRequest(first.Height)
-				peer := bcR.Switch.Peers().Get(peerID)
-				if peer != nil {
-					// NOTE: we've already removed the peer's request, but we still need
-					// to clean up the rest.
-					bcR.Switch.StopPeerForError(peer, err)
+				r.Logger.Error(
+					err.Error(),
+					"last_commit", second.LastCommit,
+					"block_id", firstID,
+					"height", first.Height,
+				)
+
+				// NOTE: We've already removed the peer's request, but we still need
+				// to clean up the rest.
+				peerID := r.pool.RedoRequest(first.Height)
+				r.blockchainCh.Error() <- p2p.PeerError{
+					PeerID:   peerID,
+					Err:      err,
+					Severity: p2p.PeerErrorSeverityLow,
 				}
 
-				peerID2 := bcR.pool.RedoRequest(second.Height)
+				peerID2 := r.pool.RedoRequest(second.Height)
 				if peerID2 != peerID {
-					if peer2 := bcR.Switch.Peers().Get(peerID2); peer2 != nil {
-						bcR.Switch.StopPeerForError(peer2, err)
+					r.blockchainCh.Error() <- p2p.PeerError{
+						PeerID:   peerID2,
+						Err:      err,
+						Severity: p2p.PeerErrorSeverityLow,
 					}
 				}
 
 				continue FOR_LOOP
 			} else {
-				bcR.pool.PopRequest()
+				r.pool.PopRequest()
 
-				// TODO: batch saves so we dont persist to disk every block
-				bcR.store.SaveBlock(first, firstParts, second.LastCommit)
+				// TODO: batch saves so we do not persist to disk every block
+				r.store.SaveBlock(first, firstParts, second.LastCommit)
 
-				// TODO: same thing for app - but we would need a way to get the hash
-				// without persisting the state.
 				var err error
-				state, _, err = bcR.blockExec.ApplyBlock(state, firstID, first)
+
+				// TODO: Same thing for app - but we would need a way to get the hash
+				// without persisting the state.
+				state, _, err = r.blockExec.ApplyBlock(state, firstID, first)
 				if err != nil {
-					// TODO This is bad, are we zombie?
-					panic(fmt.Sprintf("Failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
+					// TODO: This is bad, are we zombie?
+					panic(fmt.Sprintf("failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
 				}
+
 				blocksSynced++
 
 				if blocksSynced%100 == 0 {
 					lastRate = 0.9*lastRate + 0.1*(100/time.Since(lastHundred).Seconds())
-					bcR.Logger.Info("Fast Sync Rate",
-						"height", bcR.pool.height, "max_peer_height", bcR.pool.MaxPeerHeight(), "blocks/s", lastRate)
+					r.Logger.Info(
+						"fast sync rate",
+						"height", r.pool.height,
+						"max_peer_height", r.pool.MaxPeerHeight(),
+						"blocks/s", lastRate,
+					)
+
 					lastHundred = time.Now()
 				}
 			}
+
 			continue FOR_LOOP
 
-		case <-bcR.Quit():
+		case <-r.Quit():
```

Comment:
We should probably select on `closeCh` here instead -- see other comment.

Reply:
Yeah, we can use `closeCh` here.